File: old-legacy-code/src/trading/live_trading/continuation_modules/tick_processor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Continuation Tick Processor Module

Implements self-contained tick processing for continuation trading stocks
to eliminate cross-contamination and nested loops. Each stock processes only
its own ticks using its own price data and state-based routing.
"""

from datetime import datetime
from typing import Optional
import logging

logger = logging.getLogger(__name__)


class ContinuationTickProcessor:
    """Handles tick processing for individual continuation stocks"""
    
    def __init__(self, stock):
        """
        Initialize tick processor for a specific stock
        
        Args:
            stock: StockState instance from continuation_stock_monitor
        """
        self.stock = stock
    
    def process_tick(self, price: float, timestamp: datetime):
        """
        Process a tick for THIS stock only with real-time high/low tracking
        
        Args:
            price: Current price for THIS stock
            timestamp: Tick timestamp
        """
        
        # Always update price tracking regardless of state
        self.stock.update_price(price, timestamp)
        
        # Real-time high/low tracking for continuation stocks
        # This should be called for all active stocks to update entry levels
        if self.stock.is_active:
            self._track_entry_levels(price, timestamp)
        
        # Route to state-specific handlers
        if self.stock.is_active and self.stock.entry_ready and not self.stock.entered:
            self._handle_entry_monitoring(price, timestamp)
        
        elif self.stock.entered:
            self._handle_exit_monitoring(price, timestamp)
        
        # Other states don't need tick processing:
        # - Not active: Rejected or not monitoring
        # - Not entry ready: Waiting for entry time
        # - Already entered: Handled by exit monitoring

    def _track_entry_levels(self, price: float, timestamp: datetime):
        """
        Track entry levels for continuation stocks
        
        Args:
            price: Current price for THIS stock
            timestamp: Current timestamp
        """
        
        # Always update daily high/low tracking
        self.stock.daily_high = max(self.stock.daily_high, price)
        self.stock.daily_low = min(self.stock.daily_low, price)
        
        # For continuation stocks, entry high should be the daily high
        # This allows tracking of actual price movement during the session
        if self.stock.daily_high > 0:  # Ensure we have a valid high
            new_entry_high = self.stock.daily_high
            new_entry_sl = new_entry_high * (1 - 0.04)  # 4% SL
            
            # Update entry levels if they've changed
            if (self.stock.entry_high is None or 
                new_entry_high != self.stock.entry_high or
                new_entry_sl != self.stock.entry_sl):
                self.stock.entry_high = new_entry_high
                self.stock.entry_sl = new_entry_sl
                
                # CRITICAL FIX: Only set entry_ready = True if entry time has been reached
                # This prevents entries from happening before 11:16:00
                from config import ENTRY_TIME
                current_time = timestamp.time()
                
                if current_time >= ENTRY_TIME:
                    self.stock.entry_time_reached = True
                    self.stock.entry_ready = True
                    logger.info(f"[{self.stock.symbol}] Continuation entry updated - High: {self.stock.entry_high:.2f}, SL: {self.stock.entry_sl:.2f} (Entry time reached)")
                else:
                    # Entry time not reached yet - continue tracking but don't allow entries
                    logger.info(f"[{self.stock.symbol}] Continuation entry tracking - High: {self.stock.entry_high:.2f}, SL: {self.stock.entry_sl:.2f} (Waiting for entry time {ENTRY_TIME})")

    def _handle_entry_monitoring(self, price: float, timestamp: datetime):
        """
        Handle entry monitoring for continuation stocks
        
        Args:
            price: Current price for THIS stock
            timestamp: Current timestamp
        """
        # DEBUG: Add state validation logging
        logger.info(f"[{self.stock.symbol}] Entry monitoring - Current state: active={self.stock.is_active}, Entry ready: {self.stock.entry_ready}, Entered: {self.stock.entered}")
        
        # CRITICAL FIX: Additional entry time gate - prevent entries before ENTRY_TIME
        from config import ENTRY_TIME
        current_time = timestamp.time()
        
        if current_time < ENTRY_TIME:
            logger.info(f"[{self.stock.symbol}] Skipping entry - before entry time {ENTRY_TIME} (current: {current_time})")
            return
        
        # Only process entries if stock is in correct state and ready
        if not self.stock.is_active:
            logger.info(f"[{self.stock.symbol}] Skipping entry - not active")
            return
        
        if not self.stock.entry_ready:
            logger.info(f"[{self.stock.symbol}] Skipping entry - not entry ready")
            return
            
        if self.stock.entered:
            logger.info(f"[{self.stock.symbol}] Skipping entry - already entered")
            return

        # Continuation entry logic - Enter when price crosses above entry_high
        if self.stock.entry_high is not None and price >= self.stock.entry_high:
            # Enter position
            self.stock.enter_position(price, timestamp)
            
            logger.info(f"[{self.stock.symbol}] CONTINUATION TRIGGERED at {price:.2f} - Entered position")

    def _handle_exit_monitoring(self, price: float, timestamp: datetime):
        """
        Handle exit monitoring (trailing SL + exit signals)
        
        Args:
            price: Current price for THIS stock
            timestamp: Current timestamp
        """
        if not self.stock.entered or self.stock.entry_price is None:
            return
        
        # Calculate current profit percentage
        profit_pct = (price - self.stock.entry_price) / self.stock.entry_price
        
        # Check exit signal: SL hit
        if price <= self.stock.entry_sl:
            pnl = profit_pct * 100
            self.stock.exit_position(price, timestamp, "Stop Loss Hit")
            logger.info(f"[{self.stock.symbol}] EXIT at Rs{price:.2f}, PNL: {pnl:+.2f}%")

File: frontend/src/lib/live-trader/continuation/tick-processor.ts
import { StockStateEnum } from "../types";
import { ContinuationStockState } from "./stock-monitor";

function parseTime(t: string): number {
  const parts = t.split(":");
  return parseInt(parts[0]) * 60 + parseInt(parts[1]);
}

export class ContinuationTickProcessor {
  private stock: ContinuationStockState;

  constructor(stock: ContinuationStockState) {
    this.stock = stock;
  }

  processTick(price: number, timestamp: Date): void {
    this.stock.updatePrice(price, timestamp);

    if (this.stock.isActive) {
      this.trackEntryLevels(price, timestamp);
    }

    if (this.stock.state === StockStateEnum.WAITING_FOR_ENTRY) {
      this.handleEntryMonitoring(price, timestamp);
    } else if (this.stock.state === StockStateEnum.ENTERED) {
      this.handleExitMonitoring(price, timestamp);
    }
  }

  private trackEntryLevels(price: number, timestamp: Date): void {
    // Continuously ratchet entryHigh = dailyHigh (matches old tick_processor.py:59-96)
    if (this.stock.dailyHigh > 0) {
      const newHigh = this.stock.dailyHigh;
      const newSl = newHigh * (1 - this.stock.params.entrySlPct);
      if (this.stock.entryHigh === null || newHigh !== this.stock.entryHigh || newSl !== this.stock.entrySl) {
        this.stock.entryHigh = newHigh;
        this.stock.entrySl = newSl;
      }
    }

    // Auto-flag entryReady when entry time is reached (matches old entry_time_reached pattern)
    if (!this.stock.entryReady && !this.stock.entered) {
      const entryMinutes = this.stock.params.entryTime
        ? parseTime(this.stock.params.entryTime)
        : 9 * 60 + 20;
      const nowMinutes = timestamp.getHours() * 60 + timestamp.getMinutes();
      if (nowMinutes >= entryMinutes) {
        this.stock.entryReady = true;
      }
    }
  }

  private handleEntryMonitoring(price: number, timestamp: Date): void {
    if (!this.stock.isActive || this.stock.entered || !this.stock.entryReady) return;
    if (this.stock.entryHigh === null) return;

    // Defense-in-depth: don't enter before entry time (matches old 3-gate pattern)
    const entryMinutes = this.stock.params.entryTime
      ? parseTime(this.stock.params.entryTime)
      : 9 * 60 + 20;
    const nowMinutes = timestamp.getHours() * 60 + timestamp.getMinutes();
    if (nowMinutes < entryMinutes) return;

    if (price >= this.stock.entryHigh) {
      this.stock.enterPosition(price, timestamp);
    }
  }

  private handleExitMonitoring(price: number, timestamp: Date): void {
    if (!this.stock.entered || this.stock.entryPrice === null) return;

    // No trailing SL — entrySl stays fixed at 4% below the entry point from the window
    if (this.stock.entrySl !== null && price <= this.stock.entrySl) {
      this.stock.exitPosition(price, timestamp, "Stop Loss Hit");
    }
  }
}
